packages/core/core/src/requests/AssetRequestRunner.js

```diff
@@ -0,0 +1,155 @@
+// @flow strict-local
+import type WorkerFarm from '@parcel/workers';
+import type AssetGraph from '../AssetGraph';
+import type RequestTracker, {RequestRunnerAPI} from '../RequestTracker';
+import type {
+  AssetRequestDesc,
+  AssetRequestResult,
+  ParcelOptions,
+  TransformationOpts,
+} from '../types';
+
+import invariant from 'assert';
+import nullthrows from 'nullthrows';
+import {RequestRunner, generateRequestId} from '../RequestTracker';
+
+export type AssetRequest = {|
+  id: string,
+  +type: 'asset_request',
+  request: AssetRequestDesc,
+  result?: AssetRequestResult,
+|};
+
+export default class AssetRequestRunner extends RequestRunner<
+  AssetRequestDesc,
+  AssetRequestResult,
+> {
+  options: ParcelOptions;
+  optionsRef: number;
+  configRef: number;
+  runTransform: TransformationOpts => Promise<AssetRequestResult>;
+  assetGraph: AssetGraph;
+
+  constructor(opts: {|
+    tracker: RequestTracker,
+    options: ParcelOptions,
+    optionsRef: number,
+    configRef: number,
+    workerFarm: WorkerFarm,
+    assetGraph: AssetGraph,
+  |}) {
+    super(opts);
+    this.type = 'asset_request';
+    this.options = opts.options;
+    this.optionsRef = opts.optionsRef;
+    this.configRef = opts.configRef;
+    this.runTransform = opts.workerFarm.createHandle('runTransform');
+    this.assetGraph = opts.assetGraph;
+  }
+
+  async run(request: AssetRequestDesc, api: RequestRunnerAPI) {
+    api.invalidateOnFileUpdate(
+      await this.options.inputFS.realpath(request.filePath),
+    );
+    let start = Date.now();
+    let {assets, configRequests} = await this.runTransform({
+      request: request,
+      optionsRef: this.optionsRef,
+      configRef: this.configRef,
+    });
+
+    let time = Date.now() - start;
+    for (let asset of assets) {
+      asset.stats.time = time;
+    }
+    return {assets, configRequests};
+  }
+
+  onComplete(
+    request: AssetRequestDesc,
+    result: AssetRequestResult,
+    api: RequestRunnerAPI,
+  ) {
+    this.assetGraph.resolveAssetGroup(request, result.assets);
+
+    let {assets, configRequests} = result;
+
+    for (let asset of assets) {
+      for (let filePath of asset.includedFiles.keys()) {
+        api.invalidateOnFileUpdate(filePath);
+        api.invalidateOnFileDelete(filePath);
+      }
+    }
+
+    // TODO: this should no longer be needed once we have ConfigRequestRunner
+    let graph = this.tracker.graph;
+    let subrequestNodes = [];
+    // Add config requests
+    for (let {request, result} of configRequests) {
+      let id = generateRequestId('config_request', request);
+      let shouldSetupInvalidations =
```

Comment:
nit: inline or move closer to its use below

Reply:
This actually has to be on this line because we are checking `!graph.hasNode(id)` and then right after that ensuring the node exists. We only want to set up invalidation connections if this is the first time this request has run or it was just run because it was invalid.